app/helpers/metadata_helper.rb

```diff
@@ -219,9 +239,8 @@ def validate_metadata_csv_for_samples(samples, metadata, new_samples = false)
     sample_name_index = metadata["headers"].find_index("sample_name") || metadata["headers"].find_index("Sample Name")
 
     if extract_host_genome_from_metadata
-      host_genome_index = metadata["headers"].find_index("host_genome") || metadata["headers"].find_index("Host Genome")
-      host_genomes = HostGenome.where(name: metadata["rows"].map { |row| row[host_genome_index] }.uniq)
-                               .includes(:metadata_fields).to_a
+      host_genomes, host_genome_index, new_host_genomes =
+        find_or_create_host_genomes(metadata, allow_new_host_genomes)
```

Comment:
Update: It looks like you need a host_genome_id for each sample in the SampleUploadFlow, so you do need to create the host genomes here. This is unfortunate. (it would be a pretty big refactor to have `upload_samples_with_metadata` accept either a host_genome_id or a custom name.)

I would still create the user-defined host genome objects in a separate function (outside of `validate_metadata_csv_for_samples`). I don't think a validate function should create objects. This seems similar to the situation David was talking about where "saving" a sample shouldn't do something unexpected like kick off a pipeline run. You might even go so far as to have `create_custom_host_genomes` be a completely separate controller endpoint that you call in parallel with `validate_csv_for_new_samples` on the front-end.

Reply:
>  They are only doing validation, and they might change their metadata before the final submission

I agree, this is a weakness of this PR. I did it for the reason you saw, `host_genome_id`. Note there is no user impact of spurious host genomes because we filter them out. It would just be up to use to clean them up later. 

> I would still create the user-defined host genome objects in a separate function (outside of validate_metadata_csv_for_samples).

The reason I put the creation in the validate function is so invalid host genomes will be caught and processed by the existing validation code. What if `validate_metadata_csv_for_samples` initialized the new host genomes but did not save them? Would that be clean enough in your view?

Reply:
Made that change, have a look.